### PR TITLE
Add TypeScript types to Switcher, SwitcherDivider, SwitcherItem #13584

### LINKED ISSUE #13584: Add TypeScript types to Switcher, SwitcherDivider, SwitcherItem

### PR DESCRIPTION
**Pull request by shreeup**
Closes #13584 

Add TypeScript types to Switcher, SwitcherDivider, SwitcherItem

#### Changelog

**New**

- Added 3 new typescript equivalent of Switcher, SwitcherItem, SwitcherDivider components by executing git mv command and making required changes

**Changed**

- Updated snapshot to fix test

**Removed**

- Deleted old components written in .js

#### Testing / Reviewing

executed lint, run test suite and ensured no new failures after the commit, Added these component in UIShell storybook and verified.

### CODE REVIEW

**Comment by shreeup** [comment]
I have read the DCO document and I hereby sign the DCO.

**Comment by shreeup** [comment]
For 'DCO Assistant / DCO (pull_request_target)' step to pass, is there any action required from my end? I have already commented saying "I have read the DCO document and I hereby sign the DCO."

**Comment by shreeup** [comment]
recheck

File: packages/react/src/components/UIShell/UIShell.SideNav.stories.js [hunk 1]
```diff
@@ -574,3 +574,38 @@ export const SideNavWLargeSideNavItems = () => (
...
+        aria-valuenow={'100'}>
+        Dummy child1
+      </SwitcherItem>
+    </Switcher>
+    <SwitcherDivider role="separator" data-testid="test-id" />
+    <StoryContent />
+  </>
+);
+
+SwitcherItems.storyName = 'Switcher';
```
Comment (tw15egan) [inline packages/react/src/components/UIShell/UIShell.SideNav.stories.js]: We can delete this as these components are already used in other stories, like `HeaderWActionsAndSwitcher`

Comment (shreeup) [reply packages/react/src/components/UIShell/UIShell.SideNav.stories.js]: Deleted SwitcherItems as per review commendt

File: packages/react/src/components/UIShell/Switcher.tsx [hunk 2]
```diff
@@ -4,17 +4,46 @@
...
 
-const Switcher = React.forwardRef(function Switcher(props, forwardRef) {
-  const switcherRef = useRef(null);
-  const ref = useMergedRefs([switcherRef, forwardRef]);
+interface SwitcherProps {
+  /**
+   * Required props for accessibility label on the underlying menu
+   */
+  'aria-label'?: string;
+  'aria-labelledby'?: string;
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/Switcher.tsx]: Note that this doesn't match the proptypes directly. The proptypes specify one or the other must be specified, but not both, and not none.

https://github.com/carbon-design-system/carbon/blob/36685a0be5e62db11a60d938c76b1c49a402c99a/packages/react/src/prop-types/AriaPropTypes.js#L11-L14

It's not a blocker to merge this PR but a more thorough solution could look something like this:

```js
interface BaseSwitcherProps {
  /**
   * Optionally provide a custom class to apply to the underlying `<ul>` node
   */
  className?: string;
  /**
   * expects to receive <SwitcherItem />
   */
  children: ReactNode;
  // ... other base Switcher properties
}

interface SwitcherWithAriaLabel extends BaseSwitcherProps {
  'aria-label': string;
  'aria-labelledby'?: never;  
}

interface SwitcherWithAriaLabelledBy extends BaseSwitcherProps {
  'aria-label'?: never;
  'aria-labelledby': string;
}

// ... then used below as:

type SwitcherProps = SwitcherWithAriaLabel | SwitcherWithAriaLabelledBy;

// or I think you could maybe do this?:

const Switcher = forwardRef<HTMLUListElement, SwitcherWithAriaLabel | SwitcherWithAriaLabelledBy>(function Switcher(
```

File: packages/react/src/components/UIShell/Switcher.tsx [hunk 3]
```diff
@@ -60,40 +94,46 @@ const Switcher = React.forwardRef(function Switcher(props, forwardRef) {
...
-          expanded,
-        });
-      }
-
-      return React.cloneElement(child, {
+  const childrenWithProps = React.Children.map(children, (child, index) => {
+    // only setup click handlers if onChange event is passed
+    if (
+      React.isValidElement(child) &&
+      child.type?.toString() === 'SwitcherItem'
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/Switcher.tsx]: Is this a valid way to access the displayName?

File: packages/react/src/components/UIShell/Switcher.tsx [hunk 4]
```diff
@@ -60,40 +94,46 @@ const Switcher = React.forwardRef(function Switcher(props, forwardRef) {
...
+      ref={ref as React.RefObject<HTMLUListElement>}
+      className={className}
+      {...accessibilityLabel}>
       {childrenWithProps}
     </ul>
   );
 });
 
 Switcher.displayName = 'Switcher';
+
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/Switcher.tsx]:
```suggestion
```

File: packages/react/src/components/UIShell/SwitcherItem.tsx [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+export type SwitcherItemProps<E extends ElementType> = LinkProps<E> & {
+  /**
+   * Required props for accessibility label on the underlying item
+   */
+  'aria-label'?: string;
+  /**
+   * Required props for accessibility labelledby on the underlying item
+   */
+  'aria-labelledby'?: string;
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/SwitcherItem.tsx]: Same thing here about the required existence of one of these props.

File: packages/react/src/components/UIShell/SwitcherItem.tsx [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+   * event handlers
+   */
+  handleSwitcherItemFocus?: (event: {
+    currentIndex: number;
+    direction: number;
+  }) => void;
+  /**
+   * event handlers
+   */
+  onKeyDown?: (event: KeyboardEvent) => void;
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/SwitcherItem.tsx]: Minor nit, but could these be alphabetized so they match the order presented in the proptypes?

**Review by tay1orjones — COMMENTED** [review]
Thanks for this! Overall I think this looks good but have a few questions below

**Review by tay1orjones — APPROVED** [review]
Thank you!

**Review by alisonjoseph — APPROVED** [review]

**Pull request merged**